### PR TITLE
Add automated core blocks documentation

### PR DESCRIPTION
**Pull request by mkaz**
## Description

Adds a script to generate core block documentation reading all the blocks in the `packages/block-library` and using their `block.json` properties to generate the list.

**Done:** 
- [x] Improve supports and attributes nesting values.
- [x] Hook it up to `docs:build` command

## How has this been tested?

- Run script using: `node bin/api-docs/gen-block-lib-list.js`
- Confirm results write to: `docs/reference-guides/core-blocks.md`

## Screenshots 

- Link to [rendered view of new docs](https://github.com/WordPress/gutenberg/blob/docs/gen-blocks-list/docs/reference-guides/core-blocks.md).

## Types of changes

- New script in bin/api-docs/gen-block-lib-list.js to generate core-blocks document
- Lists all core blocks with description, category, supports, attributes

### CODE REVIEW

**Comment by Soean** [comment]
Fixes #6190
Fixes #24422

**Comment by mkaz** [comment]
@oandregal I think this is all ready.

 I added the process to the  `npm run docs:build` command by adding one more ` && node ./bin/api-docs/gen-block-lib-list.js` it is not the worst `package.json` command but we may want to think about a controller script to run other scripts to generate API docs if we add much more, which I hope we will. 

I'd like to see the block.json and theme.json documentation get generated from the schema definitions (https://github.com/WordPress/gutenberg/issues/36278 and https://github.com/WordPress/gutenberg/issues/36279) so we'll probably need to address when those go in.

**Comment by Soean** [comment]
Maybe we should add the values of the supports keys. For example the Custom HTML Block has this supports values which are missing:
```
customClassName - false
className - false
html - false
```

Should we add the type (string, boolean, number, ...) of the attributes?

It would increase the readability if the attributes and supports values are rendered as a list.

**Comment by mkaz** [comment]
@Soean yeh, I was wondering if we should include the types and values. I tried this format because rendering supports and attributes as a list would make each entry and the entire page too long and difficult to navigate.

What do you think of each block on its own page and an index page to link them all? Then the details page for each block can include more data and info.

**Comment by oandregal** [comment]
Hey, for color supports there's one tricky thing: [text & background are enabled by default](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#color) (if there's a `color` key in supports) unless the block opts-out explicitly.

**Comment by mkaz** [comment]
Good catch @oandregal - this also highlighted that other disabled supports were not shown, for example `html` is enabled by default and disabled in some blocks.

I tweaked the formatting and added the disabled values as strike outs, so it shows that they are explicitly disabled.

You can see the [new docs page rendered here](https://github.com/WordPress/gutenberg/blob/737461f9ee0e0a3caa4fccc9ed329069f18bcc7e/docs/reference-guides/core-blocks.md).

**Comment by oandregal** [comment]
I was thinking that background and text (color) should be listed in the block supports when the block has a color key and does not disable them explicitly. For example, in the [paragraph block](https://github.com/WordPress/gutenberg/blob/737461f9ee0e0a3caa4fccc9ed329069f18bcc7e/docs/reference-guides/core-blocks.md#paragraph) we should show that it supports background, text, and link for color (at the moment, we only show link).

**Comment by mkaz** [comment]
@oandregal Got it.

I updated the script to augment supports with the default colors if specified. 👍 

[Latest rendered file can be viewed here.](https://github.com/WordPress/gutenberg/blob/1a535f981aaa4e339488f4e9676477e6fabc2876/docs/reference-guides/core-blocks.md)

**Comment by gziolo** [comment]
> I tweaked the formatting and added the disabled values as strike outs, so it shows that they are explicitly disabled.

It's a bit confusing when you look at the document without the context of this PR. Should we maybe include an `*` with a legend at the bottom of the document?

**Comment by mkaz** [comment]
Good suggestion @gziolo, I added a callout at the top to explain the strikethrough. 👍

File: bin/api-docs/gen-block-lib-list.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * Reads block.json file and returns markdown formatted entry.
+ *
+ * @param {string} filename
+ *
+ * @return {string} markdown
+ */
+function readBlockJSON( filename ) {
+	const data = fs.readFileSync( filename, 'utf8' );
```
Comment (gziolo) [inline bin/api-docs/gen-block-lib-list.js]: Nitpick: any particular reason why the JSON file isn't loaded with `require` call that would parse to JSON behind the scenes?

**Review by gziolo — APPROVED** [review]
It looks like a good first step. The produced output is helpful enough to expose it to the Block Editor Handbook for initial feedback.

As a follow-up, it would be nice to create a simple unit test that uses some test fixtures to validate the output produced by `readBlockJSON`. This way, we will get a nice way to speed up iterations on the documentation's presentation layer.

File: bin/api-docs/gen-block-lib-list.js (hunk 1 above)
Comment (mkaz) [reply bin/api-docs/gen-block-lib-list.js]: TIL. Though I probably knew it at some other point too 😆

**Comment by Soean** [comment]
Link to the handbook: https://developer.wordpress.org/block-editor/reference-guides/core-blocks/

**Comment by gziolo** [comment]
Now that it is available in the Block Editor Handbook, it might be great to include a new field in `block.json`:

```json
{
    "since": "WordPress 5.9.0"
}
```

or

```json
{
    "since": "Gutenberg 10.7 (experimental)"
}
```
This would a similar role to `$schema`, but instead being for development purposes, it would exist solely for documentation purposes in WordPress core.

**Pull request merged**